Testnet/Ticketbooth/Ticketbooth/TicketContract_1_0_0.cs

```diff
@@ -0,0 +1,456 @@
+            copyOfTickets[i].Price = tickets[i].Price;
+        }
+
+        Tickets = copyOfTickets;
+        EndOfSale = endOfSale;
+
+        var show = new Show
+        {
+            Name = showName,
+            Organiser = organiser,
+            Time = time,
+            EndOfSale = endOfSale
+        };
+        Log(show);
+    }
+
+    /// <summary>
+    /// Called after the ending of a ticket sale to clear the contract ticket data
+    /// </summary>
+    public void EndSale()
+    {
+        Assert(Message.Sender == Owner, "Only contract owner can end sale");
+        Assert(EndOfSale != default(ulong), "Sale not currently in progress");
+        Assert(Block.Number >= EndOfSale, "Sale contract not fulfilled");
+
+        Tickets = ResetTickets(Tickets);
+        EndOfSale = default(ulong);
+    }
+
+    /// <summary>
+    /// Checks the availability of a seat
+    /// </summary>
+    /// <param name="seatIdentifierBytes">The serialized seat identifier</param>
+    /// <returns>Whether the seat is available</returns>
+    public bool CheckAvailability(byte[] seatIdentifierBytes)
+    {
+        Assert(SaleOpen, "Sale not open");
+
+        var ticket = SelectTicket(seatIdentifierBytes);
+
+        Assert(!IsDefaultSeat(ticket.Seat), "Seat not found");
+
+        return IsAvailable(ticket);
+    }
+
+    /// <summary>
+    /// Reserves a ticket for the callers address
+    /// </summary>
+    /// <param name="seatIdentifierBytes">The serialized seat identifier</param>
+    /// <param name="secret">The encrypted secret holding ticket ownership</param>
+    /// <returns>Whether the seat was successfully reserved</returns>
+    public void Reserve(byte[] seatIdentifierBytes, byte[] secret)
+    {
+        Reserve(seatIdentifierBytes, secret, null);
+    }
+
+    /// <summary>
+    /// Reserves a ticket for the callers address and with an identifier for the customer
+    /// </summary>
+    /// <param name="seatIdentifierBytes">The serialized seat identifier</param>
+    /// <param name="secret">The encrypted secret holding ticket ownership</param>
+    /// <param name="customerIdentifier">An encrypted verifiable identifier for the customer</param>
+    /// <returns>Whether the seat was successfully reserved</returns>
+    public void Reserve(byte[] seatIdentifierBytes, byte[] secret, byte[] customerIdentifier)
+    {
+        Assert(secret != null, "Invalid secret");
+        Assert(!RequireIdentityVerification || customerIdentifier != null, "Invalid customer identifier");
+        Assert(SaleOpen, "Sale not open");
+
+        var seat = Serializer.ToStruct<Seat>(seatIdentifierBytes);
+        var copyOfTickets = Tickets;
+        Ticket ticket = default(Ticket);
+        int ticketIndex = 0;
+        for (var i = 0; i < copyOfTickets.Length; i++)
+        {
+            if (SeatsAreEqual(copyOfTickets[i].Seat, seat))
+            {
+                ticketIndex = i;
```

Comment:
This logic is repeated a couple of times, consider refactoring to a new method `FindTicket` to make it a bit more readable.

Reply:
Made some changes to hopefully aid readability.